File: leetcode/medium/714-best-stock-transaction/iter-dp.cpp
/**
 * @file iter.cpp
 * @author Vortexx2
 * @brief Problem 714 - Best Time to Buy and Sell Stock with Txn. Fees
 * @approach Using an iterative approach we keep track of the max profit today
 * if we held stock or did not hold any stock
 * @date 12-07-2022
 *
 * Runtime - 176 ms O(n)
 * Memory Usage - 55 MB O(1)
 */
#include <iostream>
#include <map>
#include <vector>

using namespace std;

class Solution {
 public:
  int maxProfit(vector<int>& prices, int fee) {
    int cash = 0,  // max profit we could have if we didn't have a share of
                   // stock at the current time step
        hold = -prices[0];  // max profit we could have if we had a share of
                            // stock at the current time step

    for (int i = 1; i < prices.size(); i++) {
      // at current time step

      // max cash in hand without any stock occurs if -
      // we don't buy any stock today or
      // we sell stock that we held on till yesterday and pay the fee
      cash = max(cash, hold + prices[i] - fee);

      // similar logic applies to this as well
      hold = max(hold, cash - prices[i]);
    }

    return cash;
  }
};

File: leetcode/medium/714-best-stock-transaction/iter-dp.py
"""
  @author Vortexx2
  Problem 714 - Best Time to Buy and Sell Stocks with Txn. Fees

  Runtime - 1327 ms
  Memory Usage - 21.3 MB
"""
from typing import List


class Solution:
  def maxProfit(self, prices: List[int], fee: int) -> int:
    cash, hold = 0, -prices[0]

    for i in range(1, len(prices)):
      cash = max(cash, hold + prices[i] - fee)
      hold = max(hold, cash - prices[i])

    return cash
